File: examples/python/generate.py
from __future__ import annotations

import argparse
from pathlib import Path

import sentinel as S

CORPUS = [
    "the cat sat on the mat",
    "the dog ran in the park",
    "a quick brown fox jumps",
    "sentinel trains causal language models",
    "cuda packs batches for throughput",
    "tiny models learn short patterns",
]


def sibling_sbpe(checkpoint: Path) -> Path:
    return checkpoint.with_suffix(".sbpe")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("checkpoint", nargs="?", default="tiny_demo.snlm")
    parser.add_argument("prompt", nargs="?", default="the cat")
    parser.add_argument(
        "--tokenizer",
        type=Path,
        default=None,
        help="Path to .sbpe (default: sibling of checkpoint with .sbpe suffix)",
    )
    args = parser.parse_args()

    ckpt = Path(args.checkpoint)
    tok_path = args.tokenizer if args.tokenizer is not None else sibling_sbpe(ckpt)

    if ckpt.exists() or Path(str(ckpt) + ".safetensors").exists():
        weight_path = str(ckpt) if ckpt.exists() else str(ckpt) + ".safetensors"
        if not tok_path.is_file():
            raise SystemExit(f"generate: missing tokenizer at {tok_path}")
        tok = S.BPETokenizer.load_from(str(tok_path))
        model = S.LanguageModel(tok.vocab_size, 64, 64, 3e-3, 2, 4)
        if S.cuda_available():
            model.enable_cuda()
        model.load_checkpoint(weight_path)
        print(f"generate: loaded {weight_path} + {tok_path}")
    else:
        print(f"generate: no checkpoint at {ckpt} — training toy model first")
        tok = S.BPETokenizer()
        tok.train(CORPUS, vocab_size=256)
        model = S.LanguageModel(tok.vocab_size, 64, 64, 3e-3, 2, 4)
        train = S.LanguageModelDataset.build(CORPUS, tok, maximum_token_count=48, build_one_hot=False)
        if S.cuda_available():
            model.enable_cuda()
            model.set_prefer_flash_attention(True)
            model.enable_cuda_train()
        model.train(train, epochs=8, batch_size=4, log_every_epochs=4)
        model.save_checkpoint(str(ckpt), include_optimizer=False)
        tok.save(str(tok_path))

    prompt_ids = tok.encode(args.prompt)
    if not prompt_ids:
        raise SystemExit("empty prompt encoding")

    continuation = model.generate(prompt_ids, new_token_count=32, temperature=0.9, top_k=20, seed=7)
    full = list(prompt_ids) + list(continuation)
    print(f"prompt: {args.prompt}")
    print(f"output: {tok.decode(full)}")


if __name__ == "__main__":
    main()

File: examples/generate.cpp
/// Load a checkpoint + tokenizer (or train a tiny model) and generate text.
///
///   ./sentinel_generate [checkpoint.snlm] [prompt]
///
/// Expects sibling `{stem}.sbpe` when loading a checkpoint. If the checkpoint is
/// missing, trains the same toy setup as train_tiny and writes both files.

#include "NeuralNet/Cuda/CudaMatmul.hpp"
#include "NeuralNet/Data/LanguageModelDataset.hpp"
#include "NeuralNet/Network/LanguageModel.hpp"
#include "NeuralNet/Optimizers/Adam.hpp"
#include "NeuralNet/Tokenizer/BPETokenizer.hpp"

#include <cstdio>
#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

namespace {

const std::vector<std::string> kCorpus = {
    "the cat sat on the mat",
    "the dog ran in the park",
    "a quick brown fox jumps",
    "sentinel trains causal language models",
    "cuda packs batches for throughput",
    "tiny models learn short patterns",
};

LanguageModel makeTinyModel(int vocabSize) {
    return LanguageModel(vocabSize, 64, 64, Adam(3e-3f), 2, 4);
}

void maybeEnableCuda(LanguageModel& model, bool train) {
    if (!CudaMatmul::isAvailable()) return;
    model.enableCuda();
    model.setCudaPreferFlashAttention(true);
    if (train)
        model.enableCudaTrain();
}

std::string siblingSbpe(const std::string& checkpointPath) {
    return std::filesystem::path(checkpointPath).replace_extension(".sbpe").string();
}

} // namespace

int main(int argc, char** argv) {
    setvbuf(stdout, nullptr, _IONBF, 0);

    const std::string checkpointPath = argc > 1 ? argv[1] : "tiny_demo.snlm";
    const std::string promptText = argc > 2 ? argv[2] : "the cat";
    const std::string tokPath = siblingSbpe(checkpointPath);

    BPETokenizer tokenizer;
    LanguageModel model = makeTinyModel(/*vocabSize=*/256);

    if (std::filesystem::exists(checkpointPath)) {
        if (!std::filesystem::exists(tokPath)) {
            std::cerr << "generate: missing tokenizer at " << tokPath
                      << " (expected sibling .sbpe of the checkpoint)\n";
            return 1;
        }
        tokenizer.load(tokPath);
        model = makeTinyModel(tokenizer.vocabSize());
        maybeEnableCuda(model, /*train=*/false);
        model.loadCheckpoint(checkpointPath);
        std::cout << "generate: loaded " << checkpointPath << " + " << tokPath << "\n";
    } else {
        std::cout << "generate: no checkpoint at " << checkpointPath << " — training toy model first\n";
        tokenizer.train(kCorpus, 256);
        model = makeTinyModel(tokenizer.vocabSize());
        LanguageModelDataset train = LanguageModelDataset::build(kCorpus, tokenizer, 48, false);
        maybeEnableCuda(model, /*train=*/true);
        LanguageModelDataset emptyTest;
        model.train(train, emptyTest, 8, 4, 4, 1);
        model.saveCheckpoint(checkpointPath, false);
        tokenizer.save(tokPath);
    }

    const std::vector<int> promptIds = tokenizer.encode(promptText);
    if (promptIds.empty()) {
        std::cerr << "generate: empty prompt encoding\n";
        return 1;
    }

    const std::vector<int> continuation = model.generate(promptIds, /*newTokenCount=*/32, /*temperature=*/0.9f, /*topK=*/20, /*seed=*/7u);
    std::vector<int> full = promptIds;
    full.insert(full.end(), continuation.begin(), continuation.end());

    std::cout << "prompt: " << promptText << "\n";
    std::cout << "output: " << tokenizer.decode(full) << "\n";
    return 0;
}
